### PR TITLE
SAMZA-2161: Move ChangelogPartitionManager and CoordinatorStream ConfigReader to MetadataStore.

### PR DESCRIPTION
**Pull request by shanthoosh**
Currently the metadata of a samza job is stored into a kafka topic named coordinator stream.

In samza-yarn `ApplicationMaster`, the same coordinator stream is read twice as a part of the `ApplicationMaster` startup sequence. This duplicate read unnecessarily prolongs the startup time of the application master and makes the container allocation take longer than usual. This inadvertently incurs a substantial increase in input stream processing delay depending upon the size of the coordinator stream. 

To mitigate this problem, the two util classes in samza viz `ChangelogPartitionManager`, `Config` should be moved to use the MetadataStore abstraction.

### CODE REVIEW

**Comment by shanthoosh** [comment]
@prateekm  
Please take a look when you have a chance.

**Comment by shanthoosh** [comment]
@prateekm 
Apart from adding new unit-tests and fixing the existing unit-tests, I did integration testing to verify:

  * Changelog partition mapping persistence/reading is consistent with this change.
  * Configuration written to coordinator stream is correct. 
  * Checkpoint tool works fine with this patch. 

Here're the relevant test-logs: https://gist.github.com/shanthoosh/a7cadf7fd750be1f76c1eecf0714708e

**Comment by prateekm** [comment]
@dnishimura Can you take a look at this?

File: samza-core/src/main/java/org/apache/samza/coordinator/stream/messages/SetChangelogMapping.java [hunk 1]
```diff
@@ -58,7 +58,8 @@ public String getTaskName() {
     return getKey();
   }
 
-  public int getPartition() {
-    return Integer.parseInt(getMessageValue(CHANGELOG_VALUE_KEY));
+  public Integer getPartition() {
+    String changelogPartition = getMessageValue(CHANGELOG_VALUE_KEY);
+    return (changelogPartition != null) ? Integer.parseInt(changelogPartition) : null;
```
Comment (dnishimura) [inline samza-core/src/main/java/org/apache/samza/coordinator/stream/messages/SetChangelogMapping.java]: Probably safer to replace the null check with `StringUtils.isNotBlank(changelogPartition)`

File: samza-core/src/main/java/org/apache/samza/storage/ChangelogStreamManager.java [hunk 2]
```diff
@@ -28,68 +29,83 @@
...
-    for (CoordinatorStreamMessage coordinatorStreamMessage : coordinatorStreamManager.getBootstrappedStream(SetChangelogMapping.TYPE)) {
-      SetChangelogMapping changelogMapEntry = new SetChangelogMapping(coordinatorStreamMessage);
-      changelogMapping.put(new TaskName(changelogMapEntry.getTaskName()), changelogMapEntry.getPartition());
-      LOG.debug("TaskName: {} is mapped to {}", changelogMapEntry.getTaskName(), changelogMapEntry.getPartition());
-    }
+    final Map<TaskName, Integer> changelogMapping = new HashMap<>();
+    metadataStore.all().forEach((taskName, partitionIdAsBytes) -> {
+        String partitionId = valueSerde.fromBytes(partitionIdAsBytes);
+        LOG.debug("TaskName: {} is mapped to {}", taskName, partitionId);
+        if (partitionId != null) {
```
Comment (dnishimura) [inline samza-core/src/main/java/org/apache/samza/storage/ChangelogStreamManager.java]: Probably safer to replace the null check with StringUtils.isNotBlank(partitionId)

File: samza-core/src/main/scala/org/apache/samza/checkpoint/CheckpointTool.scala [hunk 3]
```diff
@@ -126,8 +127,9 @@ object CheckpointTool {
   }
 
   def apply(config: Config, offsets: TaskNameToCheckpointMap): CheckpointTool = {
-    val coordinatorStreamManager = new CoordinatorStreamManager(config, new MetricsRegistryMap())
-    new CheckpointTool(offsets, coordinatorStreamManager)
+    val metadataStore: CoordinatorStreamStore = new CoordinatorStreamStore(config, new MetricsRegistryMap())
+    metadataStore.init()
```
Comment (dnishimura) [inline samza-core/src/main/scala/org/apache/samza/checkpoint/CheckpointTool.scala]: Any reason to instantiate and init the metadata store here if it'll be just used once and closed in the `CheckpointTool#run()` method?

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 4]
```diff
@@ -258,12 +252,12 @@ public void run() {
           Thread.sleep(jobCoordinatorSleepInterval);
         } catch (InterruptedException e) {
           isInterrupted = true;
-          log.error("Interrupted in job coordinator loop", e);
+          LOG.error("Interrupted in job coordinator loop {} ", e);
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Fixed the logging message format.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 5]
```diff
@@ -258,12 +252,12 @@ public void run() {
...
           Thread.currentThread().interrupt();
         }
       }
     } catch (Throwable e) {
-      log.error("Exception thrown in the JobCoordinator loop ", e);
+      LOG.error("Exception thrown in the JobCoordinator loop {} ", e);
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Fixed.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 6]
```diff
@@ -287,18 +281,18 @@ private void onShutDown() {
...
+    LOG.info("Stopped task manager");
 
     if (jmxServer != null) {
       try {
         jmxServer.stop();
-        log.info("Stopped JMX server");
+        LOG.info("Stopped Jmx Server");
       } catch (Throwable e) {
-        log.error("Exception while stopping JMX server", e);
+        LOG.error("Exception while stopping jmx server {}", e);
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Fixed the logging line. I did not change the log message. I think this somehow got changed  while merging with master.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 7]
```diff
@@ -287,18 +281,18 @@ private void onShutDown() {
...
       systemAdmins.stop();
       containerProcessManager.stop();
-      coordinatorStreamManager.stop();
+      coordinatorStreamStore.close();
     } catch (Throwable e) {
-      log.error("Exception while stopping cluster based job coordinator", e);
+      LOG.error("Exception while stopping task manager {}", e);
     }
-    log.info("Stopped cluster based job coordinator");
+    LOG.info("Stopped task manager");
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Fixed the logging line. I did not change the log message. I think this somehow got changed  while merging with master.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 8]
```diff
@@ -287,18 +281,18 @@ private void onShutDown() {
       inputStreamRegexMonitor.ifPresent(StreamRegexMonitor::stop);
       systemAdmins.stop();
       containerProcessManager.stop();
-      coordinatorStreamManager.stop();
+      coordinatorStreamStore.close();
     } catch (Throwable e) {
-      log.error("Exception while stopping cluster based job coordinator", e);
+      LOG.error("Exception while stopping task manager {}", e);
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Fixed the logging line. I did not change the log message. I think this got introduced while merging with master.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 9]
```diff
@@ -389,19 +383,20 @@ StreamPartitionCountMonitor getPartitionMonitor() {
...
     try {
       //Read and parse the coordinator system config.
-      log.info("Parsing coordinator system config {}", coordinatorSystemEnv);
+      LOG.info("Parsing coordinator system config {}", coordinatorSystemEnv);
       coordinatorSystemConfig =
           new MapConfig(SamzaObjectMapper.getObjectMapper().readValue(coordinatorSystemEnv, Config.class));
+      LOG.info("Using the coordinator system config: {}.", coordinatorSystemConfig);
     } catch (IOException e) {
-      log.error("Error reading coordinator stream config.", e);
+      LOG.error("Exception while reading coordinator stream config {}", e);
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Fixed the logging message.

File: samza-core/src/main/java/org/apache/samza/coordinator/stream/messages/SetChangelogMapping.java (hunk 1 above)
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/coordinator/stream/messages/SetChangelogMapping.java]: Done.

File: samza-core/src/main/scala/org/apache/samza/checkpoint/CheckpointTool.scala (hunk 3 above)
Comment (shanthoosh) [reply samza-core/src/main/scala/org/apache/samza/checkpoint/CheckpointTool.scala]: Helps to build unit tests easily by injecting a mock MetadataStore as a dependency into the checkpointTool#run method.

File: samza-core/src/main/java/org/apache/samza/storage/ChangelogStreamManager.java (hunk 2 above)
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/storage/ChangelogStreamManager.java]: Done.

File: samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java [hunk 10]
```diff
@@ -356,8 +356,8 @@ private void onShutDown() {
           public void onInputStreamsChanged(Set<SystemStream> initialInputSet, Set<SystemStream> newInputStreams,
               Map<String, Pattern> regexesMonitored) {
             if (hasDurableStores) {
-              LOG.error("New input system-streams discovered. Failing the job. New input streams: {}", newInputStreams,
-                  " Existing input streams:", inputStreamsToMonitor);
+              LOG.error("New input system-streams discovered. Failing the job. New input streams: {}" +
+                  " Existing input streams:", newInputStreams, inputStreamsToMonitor);
```
Comment (dnishimura) [inline samza-core/src/main/java/org/apache/samza/clustermanager/ClusterBasedJobCoordinator.java]: Still a mismatch of 1 format params and 2 arguments.

**Pull request merged**